src/main/java/teammates/ui/controller/InstructorHomePageData.java

```diff
@@ -26,5 +34,137 @@ public String getInstructorFeedbacksPageLinkForCourse(String courseID) {
+                       delete, disabled, disabled);
+        addAttributeIf(true, delete, "onclick", "return toggleDeleteCourseConfirmation('" + courseId + "')");
+        
+        int numberOfPendingCommentsForThisCourse = numberOfPendingComments.get(courseId);
+        if (numberOfPendingCommentsForThisCourse <= 0) {
+            return Arrays.asList(enroll, view, edit,add, archive, delete);
+        } else {
+            String pendingGraphic = "<span class=\"badge\">" + numberOfPendingCommentsForThisCourse + "</span>"
+                                    + "<span class=\"glyphicon glyphicon-comment\"></span>"
+                                    + "<span class=\"glyphicon glyphicon-arrow-right\"></span>"
+                                    + "<span class=\"glyphicon glyphicon-envelope\"></span>";
+            ElementTag pending = createButton(
+                    pendingGraphic,
+                    getInstructorClearPendingCommentsLink(courseId),
+                    "Send email notification to recipients of " + numberOfPendingCommentsForThisCourse
+                        + " pending " + (numberOfPendingCommentsForThisCourse > 1 ? "comments" : "comment"));
+    
+            return Arrays.asList(enroll, view, edit,add, archive, pending, delete);
+        }
+    }
+    
+    private List<CourseTableSessionRow> createSessionRows(List<FeedbackSessionAttributes> sessions,
+            InstructorAttributes instructor, String courseId) {
+        List<CourseTableSessionRow> rows = new ArrayList<CourseTableSessionRow>();
+        int displayedStatsCount = 0;
+        
+        for (FeedbackSessionAttributes session : sessions) {
+            String name = PageData.sanitizeForHtml(session.feedbackSessionName);
+            String tooltip = PageData.getInstructorHoverMessageForFeedbackSession(session);
+            String status = PageData.getInstructorStatusForFeedbackSession(session);
+            String href = getFeedbackSessionStatsLink(session.courseId, session.feedbackSessionName);
+            
+            String recent = "";
+            if (session.isOpened() || session.isWaitingToOpen()) {
+                recent = " recent";
+            } else if (displayedStatsCount < InstructorHomePageData.MAX_CLOSED_SESSION_STATS
+                       && !TimeHelper.isOlderThanAYear(session.createdTime)) {
+                recent = " recent";
+                ++displayedStatsCount;
+            }
+            
+            String actions = "";
+            try {
+                actions = getInstructorFeedbackSessionActions(session, false, instructor,
+                        getCourseIdSectionNamesMap(sessions).get(courseId));
+            } catch (EntityDoesNotExistException e) {
+                // nothing
+            }
+            
+            rows.add(new CourseTableSessionRow(name, tooltip, status, href, recent, actions));
+        }
+        
+        return rows;
+    }
 }
```

Comment:
PageData class should store the *page data model* for the page, not other data. For example, this variable is not part of the data model for the page `HashMap<String, CourseSummaryBundle> courses` (it holds intermediate data retrieved from the logic). The data model should be something like
```
Boolean isSortable;
String sortOrder;
Boolean isNewInstructor;
...
ArrayList<CourseTable>;
String versionNumer;
```
Furthermore, it is good if this class has an `init(HashMap<String, CourseSummaryBundle>, String sortOrder)` method that takes the data retrieved from Logic and populates the page data model in one shot. That way, we can test the logic of the page data model generation by simply unit testing that method. Of course in some cases it may not be possible to retrieve all required data from the Logic before converting into the page data model.